### PR TITLE
[native_assets_builder] Add dry run for link.dart

### PR DESCRIPTION
**Pull request by dcharkes**
In order to support link hooks in Flutter (https://github.com/flutter/flutter/issues/146263), we need dry run support for link hooks.

`package:native_assets_builder`

* added `NativeAssetsBuildRunner.linkDryRun` https://github.com/dart-lang/native/issues/1134
  * `NativeAssetsBuildRunner.link` add missing `linkModePreference` param.
  * `NativeAssetsBuildRunner.dryRun` renamed to `buildDryRun`
* Link hooks that do not have any assets for them to be linked, are no longer run.
* Link hooks are no longer run in the same topological order as build hooks, they are unordered.

`package:native_assets_cli`

* Removed `HookConfig.configFile`. It's perfectly fine for the native_assets_builder to give an arbitrary file name. (It defaults to `config.json` both for link and build hooks. They are in separate directories, so no need to use a different config name.)

### CODE REVIEW

File: pkgs/native_assets_builder/lib/src/model/build_dry_run_result.dart [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:native_assets_cli/native_assets_cli_internal.dart';
+
+import '../../native_assets_builder.dart';
+
+/// Similar to a [BuildResult], but for the case of a dry run, where not assets
+/// are actually produced.
```
Comment (HosseinYousefi) [inline pkgs/native_assets_builder/lib/src/model/build_dry_run_result.dart:10]: I don't understand this sentence: "where not assets are actually produced".

**Review by HosseinYousefi — APPROVED** [review]
LGTM!

File: pkgs/native_assets_builder/lib/src/model/build_dry_run_result.dart (hunk 1 above)
Comment (dcharkes) [reply pkgs/native_assets_builder/lib/src/model/build_dry_run_result.dart]: Updated the comment.

**Comment by dcharkes** [comment]
Thanks @HosseinYousefi ! 🙏

**Pull request merged**